flang/lib/Lower/OpenMP.cpp

```diff
@@ -672,6 +672,209 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
   createBodyOfOp<omp::WsLoopOp>(wsLoopOp, converter, currentLocation, eval,
                                 &wsLoopOpClauseList, iv);
 }
+static void
+handleOmpAtomicWrite(Fortran::lower::AbstractConverter &converter,
+                     Fortran::lower::pft::Evaluation &eval,
+                     const Fortran::parser::OmpAtomicWrite &atomicWrite) {
+
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  auto currentLocation = converter.getCurrentLocation();
+  mlir::Value address;
+  // If no hint clause is specified, the effect is as if
+  // hint(omp_sync_hint_none) had been specified.
+  uint64_t hint = 0;
+  mlir::StringAttr memory_order;
+  const auto &rightHandClauseList = std::get<2>(atomicWrite.t);
+  const auto &leftHandClauseList = std::get<0>(atomicWrite.t);
+  const auto &assignmentStmtExpr =
+      std::get<Fortran::parser::Expr>(std::get<3>(atomicWrite.t).statement.t);
+  const auto &assignmentStmtVariable = std::get<Fortran::parser::Variable>(
+      std::get<3>(atomicWrite.t).statement.t);
+  Fortran::lower::StatementContext stmtCtx;
+  auto value = fir::getBase(converter.genExprValue(
+      *Fortran::semantics::GetExpr(assignmentStmtExpr), stmtCtx));
+  if (auto varDesignator = std::get_if<
+          Fortran::common::Indirection<Fortran::parser::Designator>>(
+          &assignmentStmtVariable.u)) {
+    if (const auto *name = getDesignatorNameIfDataRef(varDesignator->value())) {
+      address = converter.getSymbolAddress(*name->symbol);
+    }
+  }
+  for (const auto &clause : leftHandClauseList.v) {
+    if (auto ompClause = std::get_if<Fortran::parser::OmpClause>(&clause.u)) {
```

Comment:
Too much redundant code. Move these out to functions such as `collectHintClause` and `collectMemoryOrderClause`?

Reply:
Rethink about it and lamada may be a better solution.

Reply:
Thanks for the comments @PeixinQiao. I will refactor code.

Reply:
Hello @PeixinQiao. I got caught up elsewhere. Please find the revisions done.

Reply:
@PeixinQiao do you have further comments here?